File: frontend/src/api/packBuilder.ts
import { apiClient } from '@/api/client';

export interface PackTypeInfo {
  pack_type: string;
  name: string;
  section_count: number;
  includes_trust: boolean;
  includes_provenance: boolean;
  readiness: 'ready' | 'partial' | 'not_ready';
  readiness_score: number;
}

export interface PackSection {
  section_name: string;
  section_type: string;
  items: Record<string, unknown>[];
  completeness: number;
  notes: string | null;
}

export interface PackConformanceResult {
  profile: string;
  result: 'pass' | 'fail' | 'partial';
  score: number;
  failed_checks: Array<{ check: string; status: string; reason?: string }>;
}

export interface PackResult {
  pack_id: string;
  building_id: string;
  pack_type: string;
  pack_name: string;
  sections: PackSection[];
  total_sections: number;
  overall_completeness: number;
  includes_trust: boolean;
  includes_provenance: boolean;
  generated_at: string;
  warnings: string[];
  caveats_count: number;
  pack_version: string;
  sha256_hash: string | null;
  financials_redacted?: boolean;
  conformance?: PackConformanceResult | null;
}

export interface AvailablePacksResponse {
  building_id: string;
  packs: PackTypeInfo[];
}

export const packBuilderApi = {
  listAvailable: async (buildingId: string): Promise<AvailablePacksResponse> => {
    const response = await apiClient.get<AvailablePacksResponse>(`/buildings/${buildingId}/packs`);
    return response.data;
  },
  generate: async (
    buildingId: string,
    packType: string,
    options?: { redact_financials?: boolean },
  ): Promise<PackResult> => {
    const response = await apiClient.post<PackResult>(`/buildings/${buildingId}/packs/${packType}`, {
      ...(options?.redact_financials !== undefined && { redact_financials: options.redact_financials }),
    });
    return response.data;
  },
};

File: backend/app/schemas/pack_builder.py
"""Pydantic schemas for multi-audience pack builder."""

import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field


class PackSection(BaseModel):
    """A single section within a generated pack."""

    section_name: str
    section_type: str
    items: list[dict[str, Any]]
    completeness: float = Field(ge=0.0, le=1.0)
    notes: str | None = None

    model_config = ConfigDict(from_attributes=True)


class PackTypeInfo(BaseModel):
    """Metadata about an available pack type."""

    pack_type: str
    name: str
    section_count: int
    includes_trust: bool
    includes_provenance: bool
    readiness: str  # ready | partial | not_ready
    readiness_score: float = Field(ge=0.0, le=1.0)


class PackGenerateRequest(BaseModel):
    """Request body for generating a pack."""

    language: str = "fr"
    redact_financials: bool = False


class PackConformanceResult(BaseModel):
    """Conformance check result attached to a generated pack."""

    profile: str
    result: str  # pass | fail | partial
    score: float = Field(ge=0.0, le=1.0)
    failed_checks: list[dict[str, Any]] = []


class PackResult(BaseModel):
    """Result of a pack generation."""

    pack_id: uuid.UUID
    building_id: uuid.UUID
    pack_type: str
    pack_name: str
    sections: list[PackSection]
    total_sections: int
    overall_completeness: float
    includes_trust: bool
    includes_provenance: bool
    generated_at: datetime
    warnings: list[str]
    caveats_count: int = 0
    pack_version: str = "1.0.0"
    sha256_hash: str | None = None
    financials_redacted: bool = False
    conformance: PackConformanceResult | None = None

    model_config = ConfigDict(from_attributes=True)


class AvailablePacksResponse(BaseModel):
    """Response listing all available pack types for a building."""

    building_id: uuid.UUID
    packs: list[PackTypeInfo]
